File: src/csharp/3_StructuralPatterns/1_Adapter/WithCaching.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using MoreLinq;
using NUnit.Framework;
using static System.Console;

namespace DotNetDesignPatternDemos.Structural.Adapter.WithCaching
{
  public class Point
  {
    public int X;
    public int Y;

    public Point(int x, int y)
    {
      this.X = x;
      this.Y = y;
    }

    protected bool Equals(Point other)
    {
      return X == other.X && Y == other.Y;
    }

    public override bool Equals(object obj)
    {
      if (ReferenceEquals(null, obj)) return false;
      if (ReferenceEquals(this, obj)) return true;
      if (obj.GetType() != this.GetType()) return false;
      return Equals((Point) obj);
    }

    public override int GetHashCode()
    {
      unchecked
      {
        return (X * 397) ^ Y;
      }
    }

    public override string ToString()
    {
      return $"({X}, {Y})";
    }
  }

  public class Line
  {
    public Point Start;
    public Point End;

    public Line(Point start, Point end)
    {
      this.Start = start;
      this.End = end;
    }

    protected bool Equals(Line other)
    {
      return Equals(Start, other.Start) && Equals(End, other.End);
    }

    public override bool Equals(object obj)
    {
      if (ReferenceEquals(null, obj)) return false;
      if (ReferenceEquals(this, obj)) return true;
      if (obj.GetType() != this.GetType()) return false;
      return Equals((Line) obj);
    }

    public override int GetHashCode()
    {
      unchecked
      {
        return ((Start != null ? Start.GetHashCode() : 0) * 397) ^ (End != null ? End.GetHashCode() : 0);
      }
    }
  }

  public abstract class VectorObject : Collection<Line>
  {
  }

  public class VectorRectangle : VectorObject
  {
    public VectorRectangle(int x, int y, int width, int height)
    {
      Add(new Line(new Point(x, y), new Point(x + width, y)));
      Add(new Line(new Point(x + width, y), new Point(x + width, y + height)));
      Add(new Line(new Point(x, y), new Point(x, y + height)));
      Add(new Line(new Point(x, y + height), new Point(x + width, y + height)));
    }
  }

  public class LineToPointAdapter : IEnumerable<Point>
  {
    private static int count = 0;
    static Dictionary<int, List<Point>> cache = new Dictionary<int, List<Point>>();
    private int hash;

    public LineToPointAdapter(Line line)
    {
      hash = line.GetHashCode();
      if (cache.ContainsKey(hash)) return; // we already have it

      WriteLine($"{++count}: Generating points for line [{line.Start.X},{line.Start.Y}]-[{line.End.X},{line.End.Y}] (with caching)");
      //                                                 ^^^^

      List<Point> points = new List<Point>();

      int left = Math.Min(line.Start.X, line.End.X);
      int right = Math.Max(line.Start.X, line.End.X);
      int top = Math.Min(line.Start.Y, line.End.Y);
      int bottom = Math.Max(line.Start.Y, line.End.Y);
      int dx = right - left;
      int dy = line.End.Y - line.Start.Y;

      if (dx == 0)
      {
        for (int y = top; y <= bottom; ++y)
        {
          points.Add(new Point(left, y));
        }
      }
      else if (dy == 0)
      {
        for (int x = left; x <= right; ++x)
        {
          points.Add(new Point(x, top));
        }
      }

      cache.Add(hash, points);
    }

    public IEnumerator<Point> GetEnumerator()
    {
      return cache[hash].GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
      return GetEnumerator();
    }
  }
  

  public class Demo
  {
    private static readonly List<VectorObject> vectorObjects = new List<VectorObject>
    {
      new VectorRectangle(1, 1, 10, 10),
      new VectorRectangle(3, 3, 6, 6)
    };

    // the interface we have
    public static void DrawPoint(Point p)
    {
      Write(".");
    }

    static void Main(string[] args)
    {
      Draw();
      Draw();
    }

    private static void Draw()
    {
      foreach (var vo in vectorObjects)
      {
        foreach (var line in vo)
        {
          var adapter = new LineToPointAdapter(line);
          adapter.ForEach(DrawPoint);
        }
      }
    }
  }
}

File: src/swift/3_StructuralPatterns/1_Adapter/WithCaching.swift
import Foundation

class Point : CustomStringConvertible, Hashable
{
  var x: Int
  var y: Int

  init(_ x: Int, _ y: Int)
  {
    self.x = x
    self.y = y
  }

  public var description: String
  {
    return "(\(x), \(y))"
  }

  var hashValue: Int
  {
    return (x * 397) ^ y
  }

  static func == (lhs: Point, rhs: Point) -> Bool
  {
    return lhs.x == rhs.x && lhs.y == rhs.y
  }
}

class Line : CustomStringConvertible, Hashable
{
  var start: Point
  var end: Point

  init(_ start: Point, _ end: Point)
  {
    self.start = start
    self.end = end
  }

  var description: String
  {
    return "Line from \(start) to \(end)"
  }

  var hashValue: Int
  {
    return (start.hashValue * 397) ^ end.hashValue
  }

  static func == (lhs: Line, rhs: Line) -> Bool
  {
    return lhs.start == rhs.start && lhs.end == rhs.end
  }
}

class VectorObject : Sequence
{
  var lines = [Line]()

  func makeIterator() -> IndexingIterator<Array<Line>>
  {
    return IndexingIterator(_elements: lines)
  }
}

class VectorRectangle : VectorObject
{
  init(_ x: Int, _ y: Int, _ width: Int, _ height: Int)
  {
    super.init()
    lines.append(Line(Point(x,y), Point(x+width, y)))
    lines.append(Line(Point(x+width, y), Point(x+width, y+height)))
    lines.append(Line(Point(x,y), Point(x,y+height)))
    lines.append(Line(Point(x,y+height), Point(x+width, y+height)))
  }
}

class LineToPointAdapter : Sequence
{
  private static var count = 0
  
  // hash of line -> points for line
  static var cache = [Int: [Point]]()
  var hash: Int

  init(_ line: Line)
  {
    // if the line is cached, don't add it
    hash = line.hashValue
    if type(of: self).cache[hash] != nil { return }

    type(of: self).count += 1
    print("\(type(of: self).count): Generating points for \(line)")

    let left = Swift.min(line.start.x, line.end.x)
    let right = Swift.max(line.start.x, line.end.x)
    let top = Swift.min(line.start.y, line.end.y)
    let bottom = Swift.max(line.start.y, line.end.y)
    let dx = right - left
    let dy = line.end.y - line.start.y

    var points = [Point]()

    if dx == 0
    {
      for y in top...bottom
      {
        points.append(Point(left, y))
      }
    } else if dy == 0
    {
      for x in left...right
      {
        points.append(Point(x,top))
      }
    }

    type(of: self).cache[hash] = points
  }

  func makeIterator() -> IndexingIterator<Array<Point>>
  {
    return IndexingIterator(_elements: type(of: self).cache[hash]!)
  }
}

func drawPoint(_ p: Point)
{
  print(".", terminator: "")
}

let vectorObjects = [
  VectorRectangle(1,1,10,10),
  VectorRectangle(3,3,6,6)
]

func draw()
{
  // unfortunately, can only draw points
  for vo in vectorObjects
  {
    for line in vo
    {
      let adapter = LineToPointAdapter(line)
      adapter.forEach{ drawPoint($0) }
    }
  }
}

func main()
{
  draw()
  draw() // shows why we need caching
}

main()